lib/Sema/ConstraintSystem.cpp

```diff
@@ -4881,8 +4881,24 @@ std::string swift::describeGenericType(ValueDecl *GP, bool includeName) {
 static bool diagnoseConflictingGenericArguments(ConstraintSystem &cs,
                                                 const SolutionDiff &diff,
                                                 ArrayRef<Solution> solutions) {
-  if (!diff.overloads.empty())
-    return false;
+  // Attempt to diagnose conflicting generic arguments only when
+  // the overload choices in the solutions cannot be the cause the
+  // conflicts, by checking that all differing overloads resolve
+  // to the same opened type in all solutions.
+  for (auto &overload : diff.overloads) {
```

Comment:
Sure, I’ve updated the PR adding more code in `diagnoseConflictingGenericArguments()` to detect operators as a special case of conflicting generic arguments and use the more specific diagnostics. I followed more or less the same approach as in `diagnoseAmbiguity()` which also deals with operators. After the change, the error is now:
```swift
let foo: Int = 1
Int(foo)..<Int16.max
        `- error: binary operator '..<' cannot be applied to operands of type 'Int' and 'Int16'
```
---
The diagnostics for conflicting generic arguments were already catching some issues and returning errors not specific to operators (prior to any of my changes), for example:
```swift
Int.min..<Int16.max
       `- error: conflicting arguments to generic parameter 'Self' ('Int' vs. 'Int16')
```
After the latest commit, those errors are also diagnosed as operator errors:
```swift
Int.min..<Int16.max
       `- error: binary operator '..<' cannot be applied to operands of type 'Int' and 'Int16'
```

Reply:
I may be a little bit out of context but if we want to diagnose the operator ambiguity, should we leave `diagnoseConflictingGenericArguments` as is and defer the diagnostic to `diagnoseAmbiguity` and special handle this there?

Reply:
It just doesn't seem to me that `diagnoseConflictingGenericArguments` if we are not going to diagnose the `Self` generic param conflict, but the operator argument mismatch.

Reply:
If a diagnostic is emitted in `diagnoseConflictingGenericArguments` it triggers an early exit that prevents `diagnoseAmbiguity` from being run. So, leaving `diagnoseConflictingGenericArguments` as-is and attempting to handle it in `diagnoseAmbiguity` only would mean that any issue that gets diagnosed by the former function would get the less specific error message. 

An example where Swift was already diagnosing the operator error in `diagnoseConflictingGenericArguments`:
```
Int.min..<Int16.max
       `- error: conflicting arguments to generic parameter 'Self' ('Int' vs. 'Int16')
```
So I think making some changes to `diagnoseConflictingGenericArguments` is unavoidable.

Reply:
> It just doesn't seem to me that diagnoseConflictingGenericArguments if we are not going to diagnose the Self generic param conflict, but the operator argument mismatch.

I was thinking of the operator argument mismatch as a special case of a conflicting generic argument (where the affected function happens to be an operator), hence why I kept it as part of the function.

Reply:
I think this `conflicting` arguments diagnostic carries too little information in operator case, ambiguity error + a note per candidate seems better because it would point out the position and overload taken.

Reply:
Oh, but the diagnostic after my changes is:
```swift
let foo: Int = 1
Int(foo)..<Int16.max
        `- error: binary operator '..<' cannot be applied to operands of type 'Int' and 'Int16'
```
So after the latest commit (949d6e46235403340855ea40d4772f347572093e) it's not emitting the `conflicting arguments` error anymore, even though it's diagnosed at the `diagnoseConflictingGenericArguments()` stage. Isn't that the diagnostic we want?

Reply:
Sure, other that one or actual "ambiguity" but as I said before, we already have ways to diagnose this separately i.e. for `1 != ""`, I don't think `diagnoseConflictingGenericArguments` is the best place to do that.

Reply:
I saw how the fix is emitted in `1 != ""`. However, that works because in `1 != ""`, the compiler skips `diagnoseConflictingGenericArguments` altogether when `diff.overloads` is not empty:
```C++
// (Unmodified version)
static bool diagnoseConflictingGenericArguments(...) {
  if (!diff.overloads.empty())
    return false; // <-- Returns here in `1 != ""`, so no generic conflicting arguments diagnostics emitted

  // ...
}
```
Fine. This would also work for #72533 (`diff.overloads` is not empty). So yes, I could make changes to `diagnoseAmbiguity` instead (where operator errors are currently handled), and avoid making any changes to `diagnoseConflictingGenericArguments`.

However, **consider the case `Int.min..<Int16.max`**. There, `diff.overloads` happens to be empty -> a _"conflicting generic arguments"_ error is raised, and `diagnoseAmbiguity` is never called, so we don't get any tailored operator diagnostic if we handle it there. This already happens in main.
```C++
// (Unmodified version)
static bool diagnoseConflictingGenericArguments(...) {
  if (!diff.overloads.empty())
    return false; // <-- DOESN'T return here in `Int.min..<Int16.max`, a "conflicting arguments" diagnostic is emitted

  // ...
}
```
So either `diagnoseConflictingGenericArguments()` is modified to detect errors in operators (as I did), or we reorder stuff in `diagnoseAmbiguityWithFixes` to try to detect errors in operators before `diagnoseConflictingGenericArguments()` is even called. Are you arguing that the latter option is better?

Reply:
I saw how the fix is emitted in `1 != ""`. However, I noticed that in `1 != ""` the compiler skips `diagnoseConflictingGenericArguments` altogether as `diff.overloads` is not empty:
```C++
// (Unmodified version)
static bool diagnoseConflictingGenericArguments(...) {
  if (!diff.overloads.empty())
    return false; // <-- Returns here in `1 != ""`, so no generic conflicting arguments diagnostics emitted

  // ...
}
```
Fine. This skip allows emitting the correct diagnostic later in `diagnoseAmbiguity`. This would also work for #72533. So I could make changes to reuse the same path as in `diagnoseAmbiguity` instead (where operator errors are currently handled), and avoid making any changes to `diagnoseConflictingGenericArguments`.

However, **consider the case `Int.min..<Int16.max`**, which should emit the same diagnostic. In that case, `diff.overloads` happens to be empty, so:
```C++
// (Unmodified version)
static bool diagnoseConflictingGenericArguments(...) {
  if (!diff.overloads.empty())
    return false; // <-- DOESN'T return here in `Int.min..<Int16.max`, a "conflicting arguments" diagnostic is emitted

  // ...
}
```
 → `diagnoseConflictingGenericArguments()` is **not skipped**
 → A _"conflicting arguments to generic parameter"_ diagnostic is emitted there
 → `diagnoseConflictingGenericArguments()` returns
 → Since something was diagnosed there, `diagnoseAmbiguityWithFixes()` returns too
 → `diagnoseAmbiguity()` is never called
 → We don't get the tailored operator diagnostic
This already happens in main. I was trying to address it as well on this PR.

So (to be able to fix the `Int.min..<Int16.max` case too), either: 
- We modify `diagnoseConflictingGenericArguments()` to emit operator-specific diagnostics when needed (as I did).
- Or, we reorder/add stuff in `diagnoseAmbiguityWithFixes` to try to detect errors in operators **before** `diagnoseConflictingGenericArguments()` is called. I didn't go with this initially because I felt it'd lead to a lot of duplicated work.

Should go with the latter option then? Or am I missing something here?

Reply:
There is a third option here - we modify the solver to either emit different fixes or fix the problem in a different way. We have some logic in `repairFailures` that deals with out of order arguments to binary functions.

I consider `diagnoseConflictingGenericArguments` to be more of a workaround then an actual solution for diagnosing conflicting arguments, it's necessary because the solver currently produces fixes in a way that is hard to diagnose.

Ideally the solver should recognize a conflict and produce a single solution with a fix that would cover both `Int` and `Int16` and bind type variable that corresponds to `Self` to a placeholder type.

Reply:
In general I think it's most important to consider what is happening during solving and adjust that to make it easier to diagnose.

Reply:
I see what you mean. I'l try to think of a way of handling this during solving and update the PR. Thanks!